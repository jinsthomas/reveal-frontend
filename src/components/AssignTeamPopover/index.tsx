import { keys } from 'lodash';
import React, { MouseEvent } from 'react';
import { Button, Form, Popover, PopoverBody, PopoverHeader } from 'reactstrap';
import OrganizationSelect from '../../containers/forms/OrganizationSelect';
import { stopPropagationAndPreventDefault } from '../../helpers/utils';
import { Organization } from '../../store/ducks/opensrp/organizations';

/** Interface for AssignTeamPopover */
export interface AssignTeamPopoverProps {
  formName: string;
  isActive: boolean;
  jurisdictionId: string;
  onClearAssignmentsButtonClick: (e: MouseEvent) => void;
  onSaveAssignmentsButtonClick: (e: MouseEvent) => void;
  onToggle: () => void;
  organizationsById: { [key: string]: Organization } | null;
  planId: string;
  target: string;
}

const AssignTeamPopover = (props: AssignTeamPopoverProps) => {
  const {
    formName,
    isActive,
    jurisdictionId,
    onClearAssignmentsButtonClick,
    onSaveAssignmentsButtonClick,
    onToggle,
    organizationsById,
    planId,
    target,
  } = props;

  const organizationsArray: Organization[] | null =
    organizationsById &&
    keys(organizationsById)
      .map((o: string) => organizationsById[o])
      .filter(o => !!o);

  const organizationSelectProps = {
    jurisdictionId,
    name: formName,
    planId,
  };

  const popoverProps = {
    isOpen: isActive,
    onClick: (e: React.MouseEvent) => stopPropagationAndPreventDefault(e),
    target,
    toggle: () => onToggle(),
  };

  return (
    <Popover {...popoverProps}>
      <PopoverHeader>Select Teams to Assign</PopoverHeader>
      <PopoverBody>
        {organizationsArray ? (
          <Form name={formName}>
            <OrganizationSelect {...organizationSelectProps} />
            <Button color="default" onClick={onClearAssignmentsButtonClick} size="xs">
              Clear
            </Button>
            <Button color="primary" onClick={onSaveAssignmentsButtonClick} size="xs">
              Save
            </Button>
          </Form>
        ) : (
          <p>No teams loaded...</p>
        )}
      </PopoverBody>
    </Popover>
  );
};

export default AssignTeamPopover;
