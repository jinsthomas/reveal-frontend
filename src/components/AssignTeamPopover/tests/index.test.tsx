import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import AssignTeamPopover, { AssignTeamPopoverProps } from '..';

describe('/components/AssignTeamPopover', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const mock: any = jest.fn();
    const props: AssignTeamPopoverProps = {
      formName: 'plan-assignment-form-outpost-number-one',
      isActive: true,
      jurisdictionId: 'outpost-number-one',
      onClearAssignmentsButtonClick: mock,
      onSaveAssignmentsButtonClick: mock,
      onToggle: mock,
      organizationsById: null,
      planId: 'alpha',
      target: 'plan-assignment-outpost-number-one',
    };
    shallow(<AssignTeamPopover {...props} />);
  });

  // todo - fix this test!!!
  // it('renders correctly', () => {
  //   const mock: any = jest.fn();
  //   const props:AssignTeamPopoverProps = {
  //     formName: 'plan-assignment-form-outpost-number-one',
  //     isActive: true,
  //     jurisdictionId: 'outpost-number-one',
  //     onClearAssignmentsButtonClick: mock,
  //     onSaveAssignmentsButtonClick: mock,
  //     onToggle: mock,
  //     organizationsById: null,
  //     planId: 'alpha',
  //     target: 'plan-assignment-outpost-number-one',
  //   }

  //   // attempted fix from https://github.com/reactstrap/reactstrap/issues/773#issuecomment-373451256
  //   var div = document.createElement('div');
  //   document.body.appendChild(div);

  //   // this is breaking on mount because the underlying Popover component can't find the button
  //   const wrapper = mount((
  //     <div>
  //       <button id="plan-assignment-outpost-number-one">Target Button</button>
  //       <AssignTeamPopover {...props} />
  //     </div>
  //   ), { attachTo: div });
  // })
});
