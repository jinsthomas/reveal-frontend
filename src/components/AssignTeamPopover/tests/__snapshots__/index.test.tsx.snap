// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/AssignTeamPopover renders and passes props correctly 1`] = `
<Popover
  isOpen={true}
  onClick={[Function]}
  placement="right"
  placementPrefix="bs-popover"
  style={
    Object {
      "minWidth": "16rem",
    }
  }
  target="plan-assignment-outpost-number-one"
  toggle={[Function]}
  trigger="click"
>
  <TooltipPopoverWrapper
    autohide={false}
    className="popover show"
    delay={
      Object {
        "hide": 250,
        "show": 0,
      }
    }
    hideArrow={false}
    innerClassName="popover-inner"
    isOpen={true}
    onClick={[Function]}
    placement="right"
    placementPrefix="bs-popover"
    style={
      Object {
        "minWidth": "16rem",
      }
    }
    target="plan-assignment-outpost-number-one"
    toggle={[Function]}
    trigger="click"
  >
    <PopperContent
      boundariesElement="scrollParent"
      className="popover show"
      container="body"
      fallbackPlacement="flip"
      flip={true}
      hideArrow={false}
      isOpen={true}
      modifiers={Object {}}
      offset={0}
      placement="right"
      placementPrefix="bs-popover"
      target="plan-assignment-outpost-number-one"
    >
      <Portal
        containerInfo={
          <body>
            <div
              id="plan-assignment-outpost-number-one"
            />
            <div>
              <div
                class="popover show bs-popover-right"
                style="position: absolute; pointer-events: none; opacity: 0;"
                x-placement="right"
              >
                <div
                  aria-hidden="true"
                  class="popover-inner"
                  role="tooltip"
                  style="min-width: 16rem;"
                >
                  <h3
                    class="popover-header"
                  >
                    Select Teams to Assign
                  </h3>
                  <div
                    class="popover-body"
                  >
                    <p>
                      No teams loaded...
                    </p>
                  </div>
                </div>
                <span
                  class="arrow"
                />
              </div>
            </div>
          </body>
        }
      >
        <div>
          <Popper
            className="popover show bs-popover-right"
            component="div"
            eventsEnabled={true}
            modifiers={
              Object {
                "flip": Object {
                  "behavior": "flip",
                  "enabled": true,
                },
                "offset": Object {
                  "offset": 0,
                },
                "preventOverflow": Object {
                  "boundariesElement": "scrollParent",
                },
                "update": Object {
                  "enabled": true,
                  "fn": [Function],
                  "order": 950,
                },
              }
            }
            placement="right"
            positionFixed={false}
            x-placement="right"
          >
            <div
              className="popover show bs-popover-right"
              style={
                Object {
                  "opacity": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                }
              }
              x-placement="right"
            >
              <div
                aria-hidden={true}
                className="popover-inner"
                onClick={[Function]}
                onKeyDown={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                role="tooltip"
                style={
                  Object {
                    "minWidth": "16rem",
                  }
                }
              >
                <PopoverHeader
                  tag="h3"
                >
                  <h3
                    className="popover-header"
                  >
                    Select Teams to Assign
                  </h3>
                </PopoverHeader>
                <PopoverBody
                  tag="div"
                >
                  <div
                    className="popover-body"
                  >
                    <p>
                      No teams loaded...
                    </p>
                  </div>
                </PopoverBody>
              </div>
              <Arrow
                className="arrow"
              >
                <span
                  className="arrow"
                  style={Object {}}
                />
              </Arrow>
            </div>
          </Popper>
        </div>
      </Portal>
    </PopperContent>
  </TooltipPopoverWrapper>
</Popover>
`;

exports[`/components/AssignTeamPopover renders correctly when organizationsById is not null: button clear props 1`] = `
Object {
  "children": "Clear",
  "color": "default",
  "onClick": [MockFunction],
  "outline": true,
  "size": "xs",
  "tag": "button",
}
`;

exports[`/components/AssignTeamPopover renders correctly when organizationsById is not null: button save props 1`] = `
Object {
  "children": "Save",
  "color": "primary",
  "onClick": [MockFunction],
  "size": "xs",
  "tag": "button",
}
`;

exports[`/components/AssignTeamPopover renders correctly when organizationsById is not null: organization select props 1`] = `
Object {
  "assignments": Array [],
  "fetchAssignmentsAction": [Function],
  "fetchOrganizationsAction": [Function],
  "jurisdictionId": "outpost-number-one",
  "name": "plan-assignment-form-outpost-number-one",
  "organizations": Array [],
  "planId": "alpha",
  "resetPlanAssignmentsAction": [Function],
  "serviceClass": [Function],
  "value": Array [],
}
`;
